### DISCUSSION TITLE
Error when building OpenCV.js with threads option

### DISCUSSION THREAD

**Discussion by Treshank** [emscripten-core/emscripten#16642] [Q&A]
Ive been trying to build OpenCV.js with threads option, somehow cant seem to be able to get it to work. 
Using latest emsdk  
the command 
`python3 opencv/platforms/js/build_js.py build_wasm_simd_threads_latest\ --build_wasm --simd --threads --cmake_option="-
DOPENCV_EXTRA_MODULES_PATH=/home/t_rekt/Downloads/opencv_contrib/modules" --emscripten_dir /home/t_rekt/Downloads/emsdk/upstream/emscripten`

```wasm-ld: error: duplicate symbol: emscripten_thread_sleep
>>> defined in /home/t_rekt/Downloads/emsdk/upstream/emscripten/cache/sysroot/lib/wasm32-emscripten/libc.a(library_pthread_stub.o)
>>> defined in /home/t_rekt/Downloads/emsdk/upstream/emscripten/cache/sysroot/lib/wasm32-emscripten/libc-mt.a(library_pthread.o)

wasm-ld: error: duplicate symbol: emscripten_current_thread_process_queued_calls
>>> defined in /home/t_rekt/Downloads/emsdk/upstream/emscripten/cache/sysroot/lib/wasm32-emscripten/libc.a(library_pthread_stub.o)
>>> defined in /home/t_rekt/Downloads/emsdk/upstream/emscripten/cache/sysroot/lib/wasm32-emscripten/libc-mt.a(library_pthread.o)

wasm-ld: error: duplicate symbol: emscripten_main_browser_thread_id
>>> defined in /home/t_rekt/Downloads/emsdk/upstream/emscripten/cache/sysroot/lib/wasm32-emscripten/libc.a(pthread_self_stub.o)
>>> defined in /home/t_rekt/Downloads/emsdk/upstream/emscripten/cache/sysroot/lib/wasm32-emscripten/libc-mt.a(library_pthread.o)

wasm-ld: error: duplicate symbol: emscripten_main_thread_process_queued_calls
>>> defined in /home/t_rekt/Downloads/emsdk/upstream/emscripten/cache/sysroot/lib/wasm32-emscripten/libc.a(library_pthread_stub.o)
>>> defined in /home/t_rekt/Downloads/emsdk/upstream/emscripten/cache/sysroot/lib/wasm32-emscripten/libc-mt.a(library_pthread.o)

wasm-ld: error: duplicate symbol: __pthread_create
>>> defined in /home/t_rekt/Downloads/emsdk/upstream/emscripten/cache/sysroot/lib/wasm32-emscripten/libc.a(library_pthread_stub.o)
>>> defined in /home/t_rekt/Downloads/emsdk/upstream/emscripten/cache/sysroot/lib/wasm32-emscripten/libc-mt.a(pthread_create.o)

wasm-ld: error: duplicate symbol: __get_tp
>>> defined in /home/t_rekt/Downloads/emsdk/upstream/emscripten/cache/sysroot/lib/wasm32-emscripten/libc.a(pthread_self_stub.o)
>>> defined in /home/t_rekt/Downloads/emsdk/upstream/emscripten/cache/sysroot/lib/wasm32-emscripten/libc-mt.a(emscripten_thread_state.o)

wasm-ld: error: duplicate symbol: emscripten_is_main_runtime_thread
>>> defined in /home/t_rekt/Downloads/emsdk/upstream/emscripten/cache/sysroot/lib/wasm32-emscripten/libc.a(library_pthread_stub.o)
>>> defined in /home/t_rekt/Downloads/emsdk/upstream/emscripten/cache/sysroot/lib/wasm32-emscripten/libc-mt.a(emscripten_thread_state.o)

wasm-ld: error: duplicate symbol: __pthread_exit
>>> defined in /home/t_rekt/Downloads/emsdk/upstream/emscripten/cache/sysroot/lib/wasm32-emscripten/libc.a(library_pthread_stub.o)
>>> defined in /home/t_rekt/Downloads/emsdk/upstream/emscripten/cache/sysroot/lib/wasm32-emscripten/libc-mt.a(pthread_create.o)

wasm-ld: error: duplicate symbol: _emscripten_yield
>>> defined in /home/t_rekt/Downloads/emsdk/upstream/emscripten/cache/sysroot/lib/wasm32-emscripten/libc.a(library_pthread_stub.o)
>>> defined in /home/t_rekt/Downloads/emsdk/upstream/emscripten/cache/sysroot/lib/wasm32-emscripten/libc-mt.a(emscripten_yield.o)
em++: error: '/home/t_rekt/Downloads/emsdk/upstream/bin/wasm-ld -o ../../bin/opencv_js.wasm --gc-sections CMakeFiles/opencv_js.dir/__/js_bindings_generator/gen/bindings.cpp.o ../../lib/libopencv_core.a ../../lib/libopencv_flann.a ../../lib/libopencv_imgproc.a ../../lib/libopencv_photo.a ../../lib/libopencv_dnn.a ../../lib/libopencv_features2d.a ../../lib/libopencv_wechat_qrcode.a ../../lib/libopencv_calib3d.a ../../lib/libopencv_objdetect.a ../../lib/libopencv_video.a ../../lib/libopencv_aruco.a /home/t_rekt/Downloads/emsdk/upstream/emscripten/cache/sysroot/lib/wasm32-emscripten/libc.a ../../3rdparty/lib/libquirc.a ../../lib/libopencv_dnn.a ../../3rdparty/lib/liblibprotobuf.a ../../lib/libopencv_calib3d.a ../../lib/libopencv_features2d.a ../../lib/libopencv_flann.a ../../lib/libopencv_imgproc.a ../../lib/libopencv_core.a ../../3rdparty/lib/libzlib.a -L/home/t_rekt/Downloads/emsdk/upstream/emscripten/cache/sysroot/lib/wasm32-emscripten /home/t_rekt/Downloads/emsdk/upstream/emscripten/cache/sysroot/lib/wasm32-emscripten/crtbegin.o --whole-archive -lembind-rtti --no-whole-archive -lGL-mt -lal -lhtml5 -lstubs -lnoexit -lc-mt -ldlmalloc-mt -lcompiler_rt-mt -lc++-mt-noexcept -lc++abi-mt-noexcept -lsockets-mt -mllvm -combiner-global-alias-analysis=false -mllvm -enable-emscripten-sjlj -mllvm -disable-lsr --import-undefined --import-memory --shared-memory --strip-debug --export-if-defined=main --export-if-defined=_emscripten_thread_init --export-if-defined=_emscripten_thread_exit --export-if-defined=_emscripten_thread_crashed --export-if-defined=emscripten_tls_init --export-if-defined=pthread_self --export-if-defined=__start_em_asm --export-if-defined=__stop_em_asm --export=__cxa_demangle --export=stackSave --export=stackRestore --export=stackAlloc --export=__wasm_call_ctors --export=__errno_location --export=emscripten_dispatch_to_thread_ --export=_emscripten_thread_free_data --export=_emscripten_allow_main_runtime_queued_calls --export=emscripten_main_browser_thread_id --export=emscripten_main_thread_process_queued_calls --export=emscripten_run_in_main_runtime_thread_js --export=emscripten_stack_set_limits --export=emscripten_builtin_memalign --export=malloc --export=free --export=ntohs --export=htons --export-table -z stack-size=5242880 --initial-memory=134217728 --no-entry --max-memory=1073741824 --global-base=1024' failed (returned 1)
make[3]: *** [modules/js/CMakeFiles/opencv_js.dir/build.make:127: bin/opencv_js.js] Error 1
make[3]: Leaving directory '/home/t_rekt/Downloads/build_wasm_simd_threads_2_0_10'
make[2]: *** [CMakeFiles/Makefile2:4567: modules/js/CMakeFiles/opencv_js.dir/all] Error 2
make[2]: Leaving directory '/home/t_rekt/Downloads/build_wasm_simd_threads_2_0_10'
make[1]: *** [CMakeFiles/Makefile2:4601: modules/js/CMakeFiles/opencv.js.dir/rule] Error 2
make[1]: Leaving directory '/home/t_rekt/Downloads/build_wasm_simd_threads_2_0_10'
make: *** [Makefile:827: opencv.js] Error 2
Traceback (most recent call last):
  File "/home/t_rekt/Downloads/opencv/platforms/js/build_js.py", line 296, in <module>
    builder.build_opencvjs()
  File "/home/t_rekt/Downloads/opencv/platforms/js/build_js.py", line 203, in build_opencvjs
    execute(["make", "-j", str(multiprocessing.cpu_count()), "opencv.js"])
  File "/home/t_rekt/Downloads/opencv/platforms/js/build_js.py", line 23, in execute
    raise Fail("Child returned: %s" % retcode)
__main__.Fail: Child returned: 2
```

**Comment by sbc100** [comment]
Somehow two different versions of libc are being included.  It looks like maybe `-lc` is being injected explicitly?  Can you find out where its being added and remove it from the command line?

If you can find the failing emcc command line, it should not have `-lc` in there.

**Comment by nerdola-de-cartola** [comment]
> Ive been trying to build OpenCV.js with threads option, somehow cant seem to be able to get it to work. Using latest emsdk the command `python3 opencv/platforms/js/build_js.py build_wasm_simd_threads_latest\ --build_wasm --simd --threads --cmake_option="- DOPENCV_EXTRA_MODULES_PATH=/home/t_rekt/Downloads/opencv_contrib/modules" --emscripten_dir /home/t_rekt/Downloads/emsdk/upstream/emscripten`
> 
> ```
> >>> defined in /home/t_rekt/Downloads/emsdk/upstream/emscripten/cache/sysroot/lib/wasm32-emscripten/libc.a(library_pthread_stub.o)
> >>> defined in /home/t_rekt/Downloads/emsdk/upstream/emscripten/cache/sysroot/lib/wasm32-emscripten/libc-mt.a(library_pthread.o)
> 
> wasm-ld: error: duplicate symbol: emscripten_current_thread_process_queued_calls
> >>> defined in /home/t_rekt/Downloads/emsdk/upstream/emscripten/cache/sysroot/lib/wasm32-emscripten/libc.a(library_pthread_stub.o)
> >>> defined in /home/t_rekt/Downloads/emsdk/upstream/emscripten/cache/sysroot/lib/wasm32-emscripten/libc-mt.a(library_pthread.o)
> 
> wasm-ld: error: duplicate symbol: emscripten_main_browser_thread_id
> >>> defined in /home/t_rekt/Downloads/emsdk/upstream/emscripten/cache/sysroot/lib/wasm32-emscripten/libc.a(pthread_self_stub.o)
> >>> defined in /home/t_rekt/Downloads/emsdk/upstream/emscripten/cache/sysroot/lib/wasm32-emscripten/libc-mt.a(library_pthread.o)
> 
> wasm-ld: error: duplicate symbol: emscripten_main_thread_process_queued_calls
> >>> defined in /home/t_rekt/Downloads/emsdk/upstream/emscripten/cache/sysroot/lib/wasm32-emscripten/libc.a(library_pthread_stub.o)
> >>> defined in /home/t_rekt/Downloads/emsdk/upstream/emscripten/cache/sysroot/lib/wasm32-emscripten/libc-mt.a(library_pthread.o)
> 
> wasm-ld: error: duplicate symbol: __pthread_create
> >>> defined in /home/t_rekt/Downloads/emsdk/upstream/emscripten/cache/sysroot/lib/wasm32-emscripten/libc.a(library_pthread_stub.o)
> >>> defined in /home/t_rekt/Downloads/emsdk/upstream/emscripten/cache/sysroot/lib/wasm32-emscripten/libc-mt.a(pthread_create.o)
> 
> wasm-ld: error: duplicate symbol: __get_tp
> >>> defined in /home/t_rekt/Downloads/emsdk/upstream/emscripten/cache/sysroot/lib/wasm32-emscripten/libc.a(pthread_self_stub.o)
> >>> defined in /home/t_rekt/Downloads/emsdk/upstream/emscripten/cache/sysroot/lib/wasm32-emscripten/libc-mt.a(emscripten_thread_state.o)
> 
> wasm-ld: error: duplicate symbol: emscripten_is_main_runtime_thread
> >>> defined in /home/t_rekt/Downloads/emsdk/upstream/emscripten/cache/sysroot/lib/wasm32-emscripten/libc.a(library_pthread_stub.o)
> >>> defined in /home/t_rekt/Downloads/emsdk/upstream/emscripten/cache/sysroot/lib/wasm32-emscripten/libc-mt.a(emscripten_thread_state.o)
> 
> wasm-ld: error: duplicate symbol: __pthread_exit
> >>> defined in /home/t_rekt/Downloads/emsdk/upstream/emscripten/cache/sysroot/lib/wasm32-emscripten/libc.a(library_pthread_stub.o)
> >>> defined in /home/t_rekt/Downloads/emsdk/upstream/emscripten/cache/sysroot/lib/wasm32-emscripten/libc-mt.a(pthread_create.o)
> 
> wasm-ld: error: duplicate symbol: _emscripten_yield
> >>> defined in /home/t_rekt/Downloads/emsdk/upstream/emscripten/cache/sysroot/lib/wasm32-emscripten/libc.a(library_pthread_stub.o)
> >>> defined in /home/t_rekt/Downloads/emsdk/upstream/emscripten/cache/sysroot/lib/wasm32-emscripten/libc-mt.a(emscripten_yield.o)
> em++: error: '/home/t_rekt/Downloads/emsdk/upstream/bin/wasm-ld -o ../../bin/opencv_js.wasm --gc-sections CMakeFiles/opencv_js.dir/__/js_bindings_generator/gen/bindings.cpp.o ../../lib/libopencv_core.a ../../lib/libopencv_flann.a ../../lib/libopencv_imgproc.a ../../lib/libopencv_photo.a ../../lib/libopencv_dnn.a ../../lib/libopencv_features2d.a ../../lib/libopencv_wechat_qrcode.a ../../lib/libopencv_calib3d.a ../../lib/libopencv_objdetect.a ../../lib/libopencv_video.a ../../lib/libopencv_aruco.a /home/t_rekt/Downloads/emsdk/upstream/emscripten/cache/sysroot/lib/wasm32-emscripten/libc.a ../../3rdparty/lib/libquirc.a ../../lib/libopencv_dnn.a ../../3rdparty/lib/liblibprotobuf.a ../../lib/libopencv_calib3d.a ../../lib/libopencv_features2d.a ../../lib/libopencv_flann.a ../../lib/libopencv_imgproc.a ../../lib/libopencv_core.a ../../3rdparty/lib/libzlib.a -L/home/t_rekt/Downloads/emsdk/upstream/emscripten/cache/sysroot/lib/wasm32-emscripten /home/t_rekt/Downloads/emsdk/upstream/emscripten/cache/sysroot/lib/wasm32-emscripten/crtbegin.o --whole-archive -lembind-rtti --no-whole-archive -lGL-mt -lal -lhtml5 -lstubs -lnoexit -lc-mt -ldlmalloc-mt -lcompiler_rt-mt -lc++-mt-noexcept -lc++abi-mt-noexcept -lsockets-mt -mllvm -combiner-global-alias-analysis=false -mllvm -enable-emscripten-sjlj -mllvm -disable-lsr --import-undefined --import-memory --shared-memory --strip-debug --export-if-defined=main --export-if-defined=_emscripten_thread_init --export-if-defined=_emscripten_thread_exit --export-if-defined=_emscripten_thread_crashed --export-if-defined=emscripten_tls_init --export-if-defined=pthread_self --export-if-defined=__start_em_asm --export-if-defined=__stop_em_asm --export=__cxa_demangle --export=stackSave --export=stackRestore --export=stackAlloc --export=__wasm_call_ctors --export=__errno_location --export=emscripten_dispatch_to_thread_ --export=_emscripten_thread_free_data --export=_emscripten_allow_main_runtime_queued_calls --export=emscripten_main_browser_thread_id --export=emscripten_main_thread_process_queued_calls --export=emscripten_run_in_main_runtime_thread_js --export=emscripten_stack_set_limits --export=emscripten_builtin_memalign --export=malloc --export=free --export=ntohs --export=htons --export-table -z stack-size=5242880 --initial-memory=134217728 --no-entry --max-memory=1073741824 --global-base=1024' failed (returned 1)
> make[3]: *** [modules/js/CMakeFiles/opencv_js.dir/build.make:127: bin/opencv_js.js] Error 1
> make[3]: Leaving directory '/home/t_rekt/Downloads/build_wasm_simd_threads_2_0_10'
> make[2]: *** [CMakeFiles/Makefile2:4567: modules/js/CMakeFiles/opencv_js.dir/all] Error 2
> make[2]: Leaving directory '/home/t_rekt/Downloads/build_wasm_simd_threads_2_0_10'
> make[1]: *** [CMakeFiles/Makefile2:4601: modules/js/CMakeFiles/opencv.js.dir/rule] Error 2
> make[1]: Leaving directory '/home/t_rekt/Downloads/build_wasm_simd_threads_2_0_10'
> make: *** [Makefile:827: opencv.js] Error 2
> Traceback (most recent call last):
>   File "/home/t_rekt/Downloads/opencv/platforms/js/build_js.py", line 296, in <module>
>     builder.build_opencvjs()
>   File "/home/t_rekt/Downloads/opencv/platforms/js/build_js.py", line 203, in build_opencvjs
>     execute(["make", "-j", str(multiprocessing.cpu_count()), "opencv.js"])
>   File "/home/t_rekt/Downloads/opencv/platforms/js/build_js.py", line 23, in execute
>     raise Fail("Child returned: %s" % retcode)
> __main__.Fail: Child returned: 2
> ```

Hi I'm also trying to build opencv to a wasm version

Have you mange to do it? There are any tutorial links that I can use?

I can't find any info about that in the emscripten community

Thank you for your time
